src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,177 @@
+                return true;
+            }
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void addFirst(T value) {
+        Node<T> insertNode = new Node<>(value, head);
+        head.prev = insertNode;
+        head = insertNode;
+        size++;
+    }
+
+    private void insert(T value, int index) {
+        Node<T> oldNode = getNode(index);
+        Node<T> insertNode = new Node<>(oldNode.prev, value, oldNode);
+        oldNode.prev.next = insertNode;
+        oldNode.prev = insertNode;
+        size++;
+    }
+
+    private T unlink(Node<T> removedNode) {
+        final T element = removedNode.value;
+        final Node<T> next = removedNode.next;
+        final Node<T> prev = removedNode.prev;
+
+        if (prev == null) {
+            head = next;
+        } else {
+            prev.next = next;
+            removedNode.prev = null;
+        }
+
+        if (next == null) {
+            tail = prev;
+        } else {
+            next.prev = prev;
+            removedNode.next = null;
+        }
+
+        removedNode.value = null;
+        size--;
+        return element;
+    }
+
+    private void checkAddRange(int index) {
+        if (index > size || index < 0) {
+            throw new IndexOutOfBoundsException("Illegal index: " + index);
+        }
+    }
+
+    private void checkRange(int index) {
+        if (index >= size || index < 0) {
+            throw new IndexOutOfBoundsException("Illegal index: " + index);
+        }
+    }
+
+    private Node<T> getNode(int index) {
+        checkRange(index);
+        return (index <= (size >> 1)) ? searchFromHead(index) : searchFromTail(index);
+    }
+
+    private Node<T> searchFromHead(int index) {
+        Node<T> currentNode = head;
+        for (int i = 0; i < index; ++i) {
+            currentNode = currentNode.next;
+        }
+        return currentNode;
+    }
+
+    private Node<T> searchFromTail(int index) {
+        Node<T> currentNode = tail;
+        for (int i = size - 1; i > index; --i) {
+            currentNode = currentNode.prev;
+        }
+        return currentNode;
+    }
+
+    private boolean areEqualsValues(T first, T second) {
+        return first == second || first != null && first.equals(second);
+    }
+
+    private static class Node<T> {
+        private Node<T> prev;
+        private T value;
+        private Node<T> next;
+
+        private Node(Node<T> prev, T value, Node<T> next) {
+            this.prev = prev;
+            this.value = value;
+            this.next = next;
+        }
+
+        private Node(Node<T> prev, T value) {
+            this.prev = prev;
+            this.value = value;
+        }
+
+        private Node(T value, Node<T> next) {
+            this.value = value;
+            this.next = next;
+        }
```

Comment:
As a common mistake, when creating a new Node, pass all the fields to Node constructor

Reply:
I disagree. It is a normal practice to overload the constructor. In order not to hardcode the null value directly in the code, the following constructors are used that set the null value by default.